[WIP] Tests

tarzzz opened the pull request:
This is a work in progress [WiP] pull request to initiate writing tests for the visualization package.
Some things to note:
1. what method should be used for comparing two lists. either by matching each element explicitly.? or is there any specific method

2). Some classes may not have a point_list. ex. Sphere, or Cylinder. So it would be optional arg, in class Shape, and maybe ommited in subclasses **init**().

tarzzz committed:
Added some tests for Shape class and Cube class.

To be noted that Cube class is a subclass for Shape class
Also the method .homogeneous_transform() is not in Shape class.
It would be rather implemented in VisualizationFrame class.

chrisdembia commented:
What do you think about testing setter methods as well? Either `shape.set_color('RED')` or `shape.color = 'RED'`.

chrisdembia commented:
Maybe you can also test invalid user input. See how you might do this here: .

Also, are you sure we want Shape's to have an origin attribute? (1) the origin attribute is only meaningful/useful/not confusing if we specify in the documentaiton how the shape is drawn with respect to the origin. E.g. for a sphere you would say in the documentation 'Sphere is centered at origin'. For a sphere it's obvious, but it's not for a cube; you might choose the origin to be a vertex or the centroid of the cube; those are two valid options. (2) the location of the sphere is set by the translation of the VFrame containing the Shape, and so allowing the user to specify the origin may be redundant and thus unnecessarily complicate the interface.

tarzzz committed:
Added some tests for the existing test_shape, and test_cube.

tarzzz committed:
Added tests for class Cylinder, which is a subclass of shape.

tarzzz commented:
So, We can have origin only for specific shapes,
Also what you mentioned earlier regarding naming attributes differently,
like, Cube.points(), instead of Cube.set_point().
I feel set_point seems more obvious, as we can have set_point methods, and
also get_point methods.

A user would be confused what does a Cube.point() does.
Although documentation would tell him, but still on first seeing, it set_
and get_ would be more implicative.

Thanks

On Wed, Jul 3, 2013 at 1:39 PM, Christopher Dembia <notifications@github.com

> wrote:
> 
> Maybe you can also test invalid user input. See how you might do this
> here: .
> 
> Also, are you sure we want Shape's to have an origin attribute? (1) the
> origin attribute is only meaningful/useful/not confusing if we specify in
> the documentaiton how the shape is drawn with respect to the origin. E.g.
> for a sphere you would say in the documentation 'Sphere is centered at
> origin'. For a sphere it's obvious, but it's not for a cube; you might
> choose the origin to be a vertex or the centroid of the cube; those are two
> valid options. (2) the location of the sphere is set by the translation of
> the VFrame containing the Shape, and so allowing the user to specify the
> origin may be redundant and thus unnecessarily complicate the interface.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/PythonDynamics/pydy-viz/pull/1#issuecomment-20443560
> .

tarzzz committed:
Added test for cone Shape, initialized a new file for writing tests for
VisualizationFrame.

chrisdembia commented:
You make a good point. And I think it's fine if you go that way. Be aware though that there seems to be a convention arising with properties.

tarzzz commented:
I have made some commits with reference to yesterday's talk with Jason.
They are more tuned to what you had mentioned.

On Thu, Jul 4, 2013 at 11:34 PM, Christopher Dembia <
notifications@github.com> wrote:

> You make a good point. And I think it's fine if you go that way. Be aware
> though that there seems to be a convention arising with properties.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/PythonDynamics/pydy-viz/pull/1#issuecomment-20503118
> .

chrisdembia commented on tests/test_vframe.py:
```diff
@@ -0,0 +1,7 @@
+from sympy.physics.mechanics import *
+#from pydy-viz import VisualizationFrame
+
+def test_visualizattion_frame():
```
Small typo

tarzzz replied:
Sorry, I didnt understood this one ..

chrisdembia replied:
`test_visualization_frame` instead of `test_visualizattion_frame`; there's just an extra `t`.

chrisdembia commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,148 @@
+#from pydy-viz import Shape, Cube, 
+
+def test_shape():
+    point_list = [[1, 0, 0], [2, 3, 1], [-1, 5, -1], [-4, 2, 5]]
```
It sounds like you guys decided to have the points be lists rather than tuples ([1, 0, 0] rather than (1, 0, 0)). Why is that?

tarzzz replied:
Its because :
 1) It is more convenient to perform numpy operations, like comparing and multiplications etc. on list of lists, as they can be readily converted to numpy arrays, as compared to tuples.
 2) A list of list is analogous to a vector3 or vector4 of three.js, and hence no modification would be needed while passing data from python to three.js

moorepants replied:
Chris,

NumPy is going to be a hard dependency beause we will need the linear algebra stuff for transformations and such. I thought these points should be array_like, like in many numpy methods. numpy.array can take a list of lists, list of tuples, tuple of tuples, etc. So any array_like should be accepted.

``` python
numpy.array([[1, 2, 3], [1, 2, 3]])
numpy.array([(1, 2, 3), (1, 2, 3)])
numpy.array(((1, 2, 3), (1, 2, 3)))
```

chrisdembia commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,148 @@
+                        [-5, 5, 5], [-5, 5, -5], \
+                           [-5, -5, 5], [-5, -5, -5]]
+
+    assert_allclose(cube.point_list,point_list)
```
space after comma

chrisdembia commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,148 @@
+    shape = Shape('shape')
+    
+    shape.point_list = point_list
+    shape.color = 'RED'  #abstract method (abc package), a @property  
```
My understanding is that color is a property of Shape, but it is not an abstract method. Abstract methods are those that have a different implementation for each subclass (of Shape). Like, all shapes must have a "do_something()" method, but they each implement it differently. Do we expect different Shape subclasses to handle color differently?

tarzzz replied:
No, all shapes handle color in same way. I was not totally clear about abstraction at that point. I will remove the comment though .

chrisdembia commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,148 @@
+    assert_allclose(cube.point_list,point_list)
+    assert_allclose(cube.center, [0, 0, 0])
+    
+    assert isinstance(cube, Shape) == True
```
Don't need the `== True`.

chrisdembia commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,148 @@
+    
+    assert isinstance(cylinder, Shape) == True
+    
+def test_cone():
```
So three.js defines cylinders in a way that removes the need for a cone class (you can define cones with their cylinder; by setting the top radius to be zero; http://threejs.org/docs/58/#Reference/Extras.Geometries/CylinderGeometry). I don't like how they do that, but maybe we want to consider matching our shapes to their shapes (Geometry)? There doesn't seem to be a need for us to limit what can be drawn in threejs; right now we would not be able to draw a tapered cone, even though threejs allows it. 

This is a very minor point though, and the extra thought may not be worth the small benefit. Let me know your thoughts though.

If we have a user of pydy-viz who is also familiar with threejs, and they see that we allow drawing less shapes than threejs allows, that user might think "well why did they do that?". So maybe we should have an answer to that question; otherwise we should allow equal functionality.

tarzzz replied:
I was just focussing on providing some basic shapes for the viz part. there are a few geometries(which are in three.js) which I am ommitting from the Shape subclass. Also although we can consider cone as a three.js cylinder, with one diameter close to zero. But since we are extending the symbolics module(SymPy), wont it would be good on the symbolic viewpoint, to have cone and cylinder differently ?
I am not sure whether this part would be commited to the SymPy later on. But I am assuming that we would want to commit as much as it is acceptable to SymPy, to reside inside SymPy code only, alongwith the rest of mechanics package.

chrisdembia replied:
Okay that makes sense.

chrisdembia commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,148 @@
+    assert cube.name == 'Cube'
+    assert cube.length == 10
+    assert cube.color == 'BLUE'
+    point_list = [[5, 5, 5], [5, 5, -5], \
```
threejs allows drawing of Cube geometry, so I don't think we need to generate our own points list.
http://threejs.org/docs/58/#Reference/Extras.Geometries/CubeGeometry

tarzzz replied:
Except  mesh, every basic geometry can be drawn without an explicit point list. So should we limit this to MeshShape class only?

chrisdembia replied:
I think so.

chrisdembia commented:
Great stuff Tarun. A few more things.
1. It's interesting to introduce the 2D shapes such as the circle. I've never thought about doing something like that. I wonder what a cylinder with 0 height would be drawn like. If that would draw a circle, then Circle can be a subclass of  Cylinder, where we internally set the height as 0.0.
2. I think before you were passing in a lot of the attributes into the constructors; how come you've move away from that? I think the tests you have are good, but what do you think about, in addition, testing that the constructors work as well? I imagine some users would prefer using a constructor because it takes less lines than does defining each property separately.
3. I don't see a reason for us to restrict the colors to a few named colors. I definitely think we SHOULD allow using named colors, as I always find that very convenient. But what about allowing specification of an RGB value as well? Or an RGBA value (last element is the alpha value/opacity)? My vote is we specify color as either as a named color, or as RGB, and we have a separate attribute that is the alpha value.
4. So say I want to use a named color, but I don't know which named colors are available? How will this be documented? It's for this reason that I like enum's in C++. Python doesn't have enums, but you could look here for ideas: http://stackoverflow.com/questions/36932/how-can-i-represent-an-enum-in-python. That way, it's right in the code itself (not just in the documentation, which could become outdated) which named colors are available.

For example, you could create a very simple class:

``` python
class Color:
    RED = 1
    GREEN = 2
```

And use it like:

``` python
myShape.color = Color.RED
```

chrisdembia commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,148 @@
+    mesh_shape.origin = [0, 0, 0]
+    
+    assert mesh_shape.name == 'mesh_shape'
+    assert_allclose(mesh_shape.points,point_list)
```
space after comma

tarzzz committed:
Added tests for Shapes as per discussed to include attributes, instead of
explicit getter,setter methods

moorepants commented:
Tarun, I didn't mean for you to remove the **init** method tests for these classes when I told you about the @property decorator. I think all of the shapes should be completely defined by their **init** method but the user should be able to seemlessly change attributes of the object without get/set methods. For example:

``` python
shape = Cube(length=1.0, color='red', ...)
assert shape.color == 'red'
assert shape.color_hex == "#FF0000"
assert length == 1.0
shape.color = "green"
assert shape.color == 'green'
assert shape.color_hex == '#00FF00'
```

tarzzz committed:
Added tests for MeshShape class

tarzzz committed:
Added tests for Plane class

chrisdembia commented:
Have either of you seen python packages or modules that have a dash in their name?

moorepants commented:
module names are underscored, but _package_ names can be anything you want. I haven't seen any convention.

tarzzz committed:
Modified tests for shapes to include test for class constructors.

tarzzz committed:
Added tests for Torus class

tarzzz commented:
Regarding colors for shapes, what would be best?
Either a class . like say color.RED or something like strings, like 'RED' with hexa values mapped to a dictionary?

moorepants commented:
@tarzzz Check out matplotlib's color module for some inspiration: http://matplotlib.org/api/colors_api.html

you can define colors in multiple ways in matplotlib:

'red' common name strings
(1.0, 0.0, 0.0) rgb tuples
'#FF0000' html hex strings
'0.75' shades of gray
etc

matplotlib is bsd style license, so you can use their code if you want

chrisdembia commented:
Oh interesting; could @tarzzz use that module directly?

On Sun, Jul 7, 2013 at 4:45 AM, Jason Moore notifications@github.comwrote:

> @tarzzz https://github.com/tarzzz Check out matplotlib's color module
> for some inspiration: http://matplotlib.org/api/colors_api.html
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/PythonDynamics/pydy-viz/pull/1#issuecomment-20569554
> .

tarzzz commented:
I have put up some basic tests for VIsualizationFrame ..
Please have a look and comment on the same ...

tarzzz commented:
Also, wont it be an overkill to use matplotlib as a dependency, only for colors, also because there is not much provision of adding grayshade colors in three.js(that i know of). 
Also I think we can provide some basic implementations of colors(string,hex, and rgba), and they can be improved then, after we have the setup working, rather than working on writing a full blown colors module(the one like matplotlib).

chrisdembia commented:
Yeah it could be overkill, I agree, unless we need it for something else as well already.

chrisdembia commented on tests/test_vframe.py:
```diff
@@ -0,0 +1,55 @@
+
+def test_visualization_frame():
+	
+	point_list1 = [[2, 3, 1], [4, 6, 2], \
```
indentation?

chrisdembia commented on tests/test_vframe.py:
```diff
@@ -0,0 +1,55 @@
+    
+    assert frame1.name == 'frame1'
+    assert frame1.reference_frame == I
+    assert frame1.point == O
```
I would call this the `origin` of the visualization frame; it's more specific than just some point in the VFrame.

tarzzz replied:
Agreed

chrisdembia commented on tests/test_vframe.py:
```diff
@@ -0,0 +1,55 @@
+    assert frame1.name == 'frame1'
+    assert frame1.reference_frame == I
+    assert frame1.point == O
+    assert frame1.shape == mesh_shape1
```
What did we decide about allowing multiple shapes in one VFrame? I don't remember.

tarzzz replied:
We dont have multiple shapes.
I am using test for attribute setting method, i.e. that would be used to change that _one_ shape with some other shape. 1 VFrame would definitely take a single shape ..

chrisdembia commented:
The interface/tests for VFrame look great. I still feel that, if we expect to use this package/module for visualization of things other than multi-body systems, there should be a separate subclass of VisualizationFrame that takes a RigidBody.

It's good we test creating a VFrame with RigidBody's. BUT I still feel that these VFrame's should be constructed automatically; it doesn't make sense to me that the user has to define all the RigidBody's, and then do the same thing for VFrame's. Seems like duplicate work. Some piece of code should create VFrames and hook them up to RigidBody's, then give the user access to the VFrame's to add Shape's, etc. Let me know what you think about this. It's not clear to me yet how this would be done.

moorepants commented:
You don't have to make matplotlib a dependency. You can copy just the
source files you want for the basic color manipulation and include the
license from matplotlib with the source. Then check it all into our repo.

Jason
moorepants.info
+01 530-601-9791

On Sun, Jul 7, 2013 at 1:00 PM, Christopher Dembia <notifications@github.com

> wrote:
> 
> Yeah it could be overkill, I agree, unless we need it for something else
> as well already.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/PythonDynamics/pydy-viz/pull/1#issuecomment-20573655
> .

chrisdembia commented:
Ahh open source...

On Sun 07 Jul 2013 11:06:46 AM PDT, Jason Moore wrote:

> You don't have to make matplotlib a dependency. You can copy just the
> source files you want for the basic color manipulation and include the
> license from matplotlib with the source. Then check it all into our repo.
> 
> Jason
> moorepants.info
> +01 530-601-9791
> 
> On Sun, Jul 7, 2013 at 1:00 PM, Christopher Dembia
> <notifications@github.com
> 
> > wrote:
> > 
> > Yeah it could be overkill, I agree, unless we need it for something
> > else
> > as well already.
> > 
> > —
> > Reply to this email directly or view it on
> > GitHubhttps://github.com/PythonDynamics/pydy-viz/pull/1#issuecomment-20573655
> > 
> > .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/PythonDynamics/pydy-viz/pull/1#issuecomment-20574641.

tarzzz committed:
Added tests for Tube class

tarzzz committed:
Added tests for Visualization frame.
constructors using a rframe/point and using rigidbody are incorporated in
the tests

tarzzz commented:
Regarding that we have to define a full rbody for the VFrame, it would be something like this ..
1) If a user is already working with rbodies, its simple .  do  .. 
  frame1 =  VFrame('frame1', rbody,shape=some_shape)
2) If a user doesnt wants to deal with rbodies, 
  frame1 = VFrame('frame1',[rframe,origin],shape=some_shape)
Our **init** methods would parse the second arg to check whether user has supplied, 
1) a rbody, 2) a rframe/point tuple,list etc. 3) a rframe/Particle list tuple etc. 

In any case finally all we need is a rframe/point combination, and I feel that it is a necessity(to have a rframe/point) to define a VFrame .

Regarding how to attach rbody data for animation purposes. we would have a generic method.
VFrame.add_simulation _data(), which would attach a JSON file containing simulation data to that frame.
This is when we dont have a rigidbody to provide, and hence _might_ work for other systems.

I would again mention that the rbody in the constructor here is _only_ for providing rframe/point combination, and It is totally optional to use it or else a rframe/point combination explicitly as I did, or a rframe/Particle combination, that I am writing tests for now ..

I hope that clears this bit. .

chrisdembia commented:
I see what you're saying but I don't think it solves our desire for modularity, if I understand that correctly. By modularity, my understanding is we want people to be able to use this package independently of pydy. In the case that someone is using the package independently of pydy, then they don't know anything about rigidbody's, and none of the classes they use should know anything about rigidbody's (even as an optional argument. imagine reading the docstring and seeing 'argument can be a rigidbody', but not even knowing what that is). But maybe I misunderstand our desire for modularity. I am not too clear on the VFrame.add_simulation_data() / JSON stuff.

tarzzz commented:
A complete cutoff from pydy would be difficult, as we will have to redesign our VFRame, to function without rframes and points,  moreover by extent of modularity, as I understand, It should be like we are able to see our visualizations without using the pydy module and mechanics methods(i.e. rbodies, Shapes etc.)

That can be achieved if a user is able to output the desired JSON(*in the relevant format) which is input to our Javascripts, by using any other module or even programming language he wants/ or maybe he justs write that JSON and then provide data from somewhere else. 
Anyways depends on what extent of modularity we want.

moorepants commented:
I think the JS side should be "modular" in the sense that if you provide the right numerical and meta data you'll get to animate things. But the Python side can be less "modular" for general use. We are in a sense writing the plugin for pydy.

moorepants commented on tests/test_shapes.py:
```diff
@@ -0,0 +1,248 @@
+    assert isinstance(circle, Shape)
+
+def test_mesh_shape():
+    point_list = [[2, 3, 1], [4, 6, 2], \
```
Please test passing in numpy arrays and various other array_like things.

tarzzz commented:
Also regarding the color, we can simply provide a string of color from a standard as defined in matplotlib's color module, and then we can utilize ColorConverter from the matplotlob module to convert those string colors to RGB and RGBA.
All the colors would ultimately be converted into RGB, RGBA before passing to JS.
So there is not much modifications to the tests. 
Also Do I would require to write tests for colors module as welll? If we are taking it from matplotlib and adding to our repo?

moorepants commented:
If you take the color converter module, you should also take the tests for the color converter module.

moorepants commented:
Moved to #3.
